pkg/nsx/services/expand.go

```diff
@@ -0,0 +1,308 @@
+package services
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/data"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	v1 "k8s.io/api/core/v1"
+	meta1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	util2 "github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+	"github.com/vmware-tanzu/nsx-operator/pkg/util"
+)
+
+// When a rule contains named port, we should consider whether the rule should be expanded to
+// multiple rules if the port name maps to conflicted port numbers.
+func (service *SecurityPolicyService) expandRule(
+	obj *v1alpha1.SecurityPolicy,
+	rule *v1alpha1.SecurityPolicyRule,
+	ruleIdx int,
+) ([]*model.Group, []*model.Rule, error) {
+	var nsxRules []*model.Rule
+	var nsxGroups []*model.Group
+
+	if len(rule.Ports) == 0 {
+		rs, err := service.buildRuleBasicInfo(obj, rule, ruleIdx, 0, 0)
+		if err != nil {
+			return nil, nil, err
+		}
+		nsxRules = append(nsxRules, rs)
+	}
+	for portIdx, port := range rule.Ports {
+		gs, rs, err := service.expandRuleByPort(obj, rule, ruleIdx, port, portIdx)
```

Comment:
Please don't use ns/gs/rs/... as the variable names, it's too hard to understand the logic!
And the naming styles, `nsxRules` and `rs` are not consistent.

Reply:
ok